skopt/searchcv.py

```diff
@@ -0,0 +1,547 @@
+        else:
+            self.optimizer_kwargs = optimizer_kwargs
+
+        # this dict is used in order to keep track of skopt Optimizer
+        # instances for different search spaces. `str(space)` is used as key
+        # as `space` is a dict, which is unhashable. however, str(space)
+        # is fixed and unique if space is not changed.
+        self.optimizer_ = {}
+        self.cv_results_ = defaultdict(list)
+
+        self.best_index_ = None
+
+        super(BayesSearchCV, self).__init__(
+             estimator=estimator, scoring=scoring, fit_params=fit_params,
+             n_jobs=n_jobs, iid=iid, refit=refit, cv=cv, verbose=verbose,
+             pre_dispatch=pre_dispatch, error_score=error_score,
+             return_train_score=return_train_score)
+
+    def _check_search_space(self, search_space):
+        """Checks whether the search space argument is correct"""
+
+        # check if space is a single dict, convert to list if so
+        if isinstance(search_space, dict):
+            search_space = [search_space]
+
+        # check if the structure of the space is proper
+        if isinstance(search_space, list):
+            # convert to just a list of dicts
+            dicts_only = []
+
+            # 1. check the case when a tuple of space, n_iter is provided
+            for elem in search_space:
+                if isinstance(elem, tuple):
+                    if len(elem) != 2:
+                        raise ValueError(
+                            "All tuples in list of search spaces should have"
+                            "length 2, and contain (dict, int), got %s" % elem
+                        )
+                    subspace, n_iter = elem
+
+                    if (not isinstance(n_iter, int)) or n_iter < 0:
+                        raise ValueError(
+                            "Number of iterations in search space should be"
+                            "positive integer, got %s in tuple %s " %
+                            (n_iter, elem)
+                        )
+
+                    # save subspaces here for further checking
+                    dicts_only.append(subspace)
+                elif isinstance(elem, dict):
+                    dicts_only.append(elem)
+                else:
+                    raise TypeError(
+                        "A search space should be provided as a dict or"
+                        "tuple (dict, int), got %s" % elem)
+
+            # 2. check all the dicts for correctness of contents
+            for subspace in dicts_only:
+                for k, v in subspace.items():
+                    if not isinstance(v, Dimension):
+                        raise TypeError(
+                            "Parameter of space should be an instance of"
+                            "skopt.space.Dimension (Real, Integer, ...),"
+                            "but in subspace %s the dimension %s is %s" %
+                            (subspace, k, v)
```

Comment:
Could we instead use `skopt.space.check_dimension`? In this way, we would have a consistent API with `Optimizer/*_minimize` in terms of how dimensions are specified. (Also, this would allow much less boilerplate code for the users.)